File: go/decode_test.go
package afi

import (
	"errors"
	"math/big"
	"strings"
	"testing"

	"github.com/ethereum/go-ethereum/accounts/abi"
	"github.com/ethereum/go-ethereum/common"
	"github.com/ethereum/go-ethereum/crypto"
)

func encodeRevert(t *testing.T, signature string, paramTypes []string, values []interface{}) []byte {
	t.Helper()
	hash := crypto.Keccak256([]byte(signature))
	selector := hash[:4]
	if len(paramTypes) == 0 {
		return selector
	}
	args := abi.Arguments{}
	for _, ts := range paramTypes {
		ty, err := abi.NewType(ts, "", nil)
		if err != nil {
			t.Fatalf("type %q: %v", ts, err)
		}
		args = append(args, abi.Argument{Type: ty})
	}
	payload, err := args.Pack(values...)
	if err != nil {
		t.Fatalf("pack: %v", err)
	}
	return append(selector, payload...)
}

func TestDecodeRevertReason_BuiltInErrorString(t *testing.T) {
	data := encodeRevert(t, "Error(string)", []string{"string"}, []interface{}{"boom"})
	r := DecodeRevertReason(data)
	if r == nil {
		t.Fatal("expected non-nil")
	}
	if r.Name != "Error" {
		t.Errorf("Name = %q", r.Name)
	}
	if r.Args[0].(string) != "boom" {
		t.Errorf("arg mismatch: %v", r.Args[0])
	}
}

func TestDecodeRevertReason_BuiltInPanic(t *testing.T) {
	data := encodeRevert(t, "Panic(uint256)", []string{"uint256"}, []interface{}{big.NewInt(0x11)})
	r := DecodeRevertReason(data)
	if r == nil || r.Name != "Panic" {
		t.Fatalf("expected Panic, got %+v", r)
	}
}

func TestDecodeRevertReason_AFIInsufficientFunds(t *testing.T) {
	data := encodeRevert(t, "InsufficientFunds(uint256)", []string{"uint256"}, []interface{}{big.NewInt(1234)})
	r := DecodeRevertReason(data)
	if r == nil {
		t.Fatal("expected non-nil")
	}
	if r.Name != "InsufficientFunds" {
		t.Errorf("Name = %q", r.Name)
	}
	if r.Args[0].(*big.Int).Cmp(big.NewInt(1234)) != 0 {
		t.Errorf("arg mismatch: %v", r.Args[0])
	}
}

func TestDecodeRevertReason_AFIZeroAddress(t *testing.T) {
	data := encodeRevert(t, "ZeroAddress()", []string{}, []interface{}{})
	r := DecodeRevertReason(data)
	if r == nil || r.Name != "ZeroAddress" {
		t.Fatalf("got %+v", r)
	}
}

func TestDecodeRevertReason_AFIDifferentAssets(t *testing.T) {
	expected := common.HexToAddress("0x4200000000000000000000000000000000000006")
	actual := common.HexToAddress("0x833589fcd6edb6e08f4c7c32d4f71b54bda02913")
	data := encodeRevert(t, "DifferentAssets(address,address)", []string{"address", "address"},
		[]interface{}{expected, actual})
	r := DecodeRevertReason(data)
	if r == nil {
		t.Fatal("expected non-nil")
	}
	if r.Name != "DifferentAssets" {
		t.Errorf("Name = %q", r.Name)
	}
	if r.Args[0].(common.Address) != expected {
		t.Errorf("expected mismatch: %v", r.Args[0])
	}
}

func TestDecodeRevertReason_OZErrors(t *testing.T) {
	cases := []string{
		"OwnableInvalidOwner(address)",
		"OwnableUnauthorizedAccount(address)",
		"ReentrancyGuardReentrantCall()",
	}
	for _, sig := range cases {
		t.Run(sig, func(t *testing.T) {
			var types []string
			var values []interface{}
			if strings.Contains(sig, "address") {
				types = []string{"address"}
				values = []interface{}{common.HexToAddress("0xdeadbeef00000000000000000000000000000000")}
			}
			data := encodeRevert(t, sig, types, values)
			r := DecodeRevertReason(data)
			if r == nil {
				t.Fatalf("expected non-nil for %s", sig)
			}
		})
	}
}

func TestDecodeRevertReason_EmptyOrShort(t *testing.T) {
	if DecodeRevertReason(nil) != nil || DecodeRevertReason([]byte{0x01, 0x02}) != nil {
		t.Error("expected nil for empty/short data")
	}
}

func TestDecodeRevertReason_UnknownSelector(t *testing.T) {
	data := encodeRevert(t, "Mystery(uint256)", []string{"uint256"}, []interface{}{big.NewInt(1)})
	if DecodeRevertReason(data) != nil {
		t.Error("expected nil for unknown selector")
	}
}

func TestRegisterCustomErrors_DecodesAfterRegistration(t *testing.T) {
	data := encodeRevert(t, "MyContractError(uint256,string)", []string{"uint256", "string"},
		[]interface{}{big.NewInt(42), "details"})

	if DecodeRevertReason(data) != nil {
		t.Fatal("should not decode before registration")
	}

	a, err := abi.JSON(strings.NewReader(`[{"type":"error","name":"MyContractError","inputs":[
		{"name":"code","type":"uint256"},
		{"name":"msg","type":"string"}
	]}]`))
	if err != nil {
		t.Fatal(err)
	}
	RegisterCustomErrors(a)

	r := DecodeRevertReason(data)
	if r == nil || r.Name != "MyContractError" {
		t.Fatalf("expected MyContractError, got %+v", r)
	}
	if r.Args[0].(*big.Int).Cmp(big.NewInt(42)) != 0 {
		t.Errorf("arg 0 mismatch: %v", r.Args[0])
	}
	if r.Args[1].(string) != "details" {
		t.Errorf("arg 1 mismatch: %v", r.Args[1])
	}
}

func TestDecodedRevert_StringFormatting(t *testing.T) {
	d := &DecodedRevert{Name: "InsufficientFunds", Args: []interface{}{big.NewInt(100)}}
	if d.String() != "InsufficientFunds(100)" {
		t.Errorf("got %q", d.String())
	}
	d0 := &DecodedRevert{Name: "ZeroAddress"}
	if d0.String() != "ZeroAddress" {
		t.Errorf("got %q", d0.String())
	}
	var nilD *DecodedRevert
	if nilD.String() != "unknown revert" {
		t.Errorf("got %q", nilD.String())
	}
}

func TestErrSimulationDecoded_Roundtrip(t *testing.T) {
	d := &DecodedRevert{Name: "InsufficientFunds", Args: []interface{}{big.NewInt(100)}}
	err := ErrSimulationDecoded(d.String(), d)
	var afiErr *AfiError
	if !errors.As(err, &afiErr) {
		t.Fatal("expected *AfiError")
	}
	if afiErr.Code != "SIMULATION_FAILED" {
		t.Errorf("Code = %q", afiErr.Code)
	}
	if afiErr.Decoded == nil || afiErr.Decoded.Name != "InsufficientFunds" {
		t.Errorf("Decoded missing: %+v", afiErr.Decoded)
	}
}

func TestGetRegisteredErrorSelectors_IncludesAFIErrors(t *testing.T) {
	selectors := GetRegisteredErrorSelectors()
	if len(selectors) < 9 {
		t.Errorf("expected at least 9 registered selectors (AFI errors), got %d", len(selectors))
	}
}

File: nodejs/src/__tests__/decode.test.ts
import { describe, expect, it } from "vitest"
import { encodeAbiParameters, toFunctionSelector } from "viem"
import {
  decodeRevertReason,
  describeDecodedRevert,
  ERROR_STRING_SELECTOR_HEX,
  getRegisteredErrors,
  PANIC_SELECTOR_HEX,
  registerCustomErrors,
} from "../decode.js"

function encodeRevert(signature: string, paramTypes: string[], values: readonly unknown[]): `0x${string}` {
  const selector = toFunctionSelector(signature)
  if (paramTypes.length === 0) return selector
  const payload = encodeAbiParameters(
    paramTypes.map((type) => ({ type })),
    values as never[],
  )
  return (selector + payload.slice(2)) as `0x${string}`
}

describe("decodeRevertReason — built-in revert", () => {
  it("decodes Error(string) reverts", () => {
    const data = encodeRevert("Error(string)", ["string"], ["boom"])
    const r = decodeRevertReason(data)
    expect(r).not.toBeNull()
    expect(r!.name).toBe("Error")
    expect(r!.args[0]).toBe("boom")
    expect(ERROR_STRING_SELECTOR_HEX).toBe("0x08c379a0")
  })

  it("decodes Panic(uint256)", () => {
    const data = encodeRevert("Panic(uint256)", ["uint256"], [0x11n])
    const r = decodeRevertReason(data)
    expect(r!.name).toBe("Panic")
    expect(r!.args[0]).toBe(0x11n)
    expect(PANIC_SELECTOR_HEX).toBe("0x4e487b71")
  })
})

describe("decodeRevertReason — AFI custom errors", () => {
  it("decodes InsufficientFunds(uint256)", () => {
    const data = encodeRevert("InsufficientFunds(uint256)", ["uint256"], [1234n])
    const r = decodeRevertReason(data)
    expect(r).not.toBeNull()
    expect(r!.name).toBe("InsufficientFunds")
    expect(r!.signature).toBe("InsufficientFunds(uint256)")
    expect(r!.args[0]).toBe(1234n)
  })

  it("decodes ZeroAddress() with no args", () => {
    const data = encodeRevert("ZeroAddress()", [], [])
    const r = decodeRevertReason(data)
    expect(r!.name).toBe("ZeroAddress")
    expect(r!.args).toHaveLength(0)
  })

  it("decodes DifferentAssets(address,address)", () => {
    const expected = "0x4200000000000000000000000000000000000006"
    const actual = "0x833589fcd6edb6e08f4c7c32d4f71b54bda02913"
    const data = encodeRevert("DifferentAssets(address,address)", ["address", "address"], [expected, actual])
    const r = decodeRevertReason(data)
    expect(r!.name).toBe("DifferentAssets")
    expect((r!.args[0] as string).toLowerCase()).toBe(expected)
    expect((r!.args[1] as string).toLowerCase()).toBe(actual)
  })

  it("decodes ReentrancyGuardReentrantCall (OpenZeppelin)", () => {
    const data = encodeRevert("ReentrancyGuardReentrantCall()", [], [])
    expect(decodeRevertReason(data)!.name).toBe("ReentrancyGuardReentrantCall")
  })

  it("decodes OwnableUnauthorizedAccount(address)", () => {
    const data = encodeRevert("OwnableUnauthorizedAccount(address)", ["address"], ["0xdeadbeef00000000000000000000000000000000"])
    expect(decodeRevertReason(data)!.name).toBe("OwnableUnauthorizedAccount")
  })
})

describe("decodeRevertReason — fallbacks", () => {
  it("returns null for empty or too-short data", () => {
    expect(decodeRevertReason(undefined)).toBeNull()
    expect(decodeRevertReason(null)).toBeNull()
    expect(decodeRevertReason("0x")).toBeNull()
    expect(decodeRevertReason("0x1234")).toBeNull()
  })

  it("returns null for unknown selectors", () => {
    const data = encodeRevert("Mystery(uint256)", ["uint256"], [1n])
    expect(decodeRevertReason(data)).toBeNull()
  })

  it("accepts hex data without the 0x prefix", () => {
    const data = encodeRevert("Error(string)", ["string"], ["boom"])
    const r = decodeRevertReason(data.slice(2)) // strip "0x"
    expect(r).not.toBeNull()
    expect(r!.name).toBe("Error")
    expect(r!.args[0]).toBe("boom")
  })

  it("returns null for non-string data", () => {
    // @ts-expect-error exercising the runtime guard
    expect(decodeRevertReason(12345)).toBeNull()
  })
})

describe("registerCustomErrors + getRegisteredErrors", () => {
  it("ignores non-error ABI entries when registering", () => {
    const before = getRegisteredErrors().length
    registerCustomErrors([
      { type: "function", name: "ignored", inputs: [] },
      { type: "event", name: "alsoIgnored", inputs: [] },
    ])
    expect(getRegisteredErrors().length).toBe(before)
  })

  it("decodes user-registered errors after registration", () => {
    const data = encodeRevert("MyContractError(uint256,string)", ["uint256", "string"], [42n, "details"])

    expect(decodeRevertReason(data)).toBeNull()

    registerCustomErrors([
      {
        type: "error",
        name: "MyContractError",
        inputs: [
          { name: "code", type: "uint256" },
          { name: "msg", type: "string" },
        ],
      },
    ])

    const r = decodeRevertReason(data)
    expect(r).not.toBeNull()
    expect(r!.name).toBe("MyContractError")
    expect(r!.args[0]).toBe(42n)
    expect(r!.args[1]).toBe("details")
  })
})

describe("describeDecodedRevert", () => {
  it("formats with args", () => {
    expect(describeDecodedRevert({ name: "InsufficientFunds", signature: "InsufficientFunds(uint256)", args: [100n] }))
      .toBe("InsufficientFunds(100)")
  })

  it("formats no-arg errors as just the name", () => {
    expect(describeDecodedRevert({ name: "ZeroAddress", signature: "ZeroAddress()", args: [] }))
      .toBe("ZeroAddress")
  })

  it("stringifies mixed bigint and non-bigint args", () => {
    expect(describeDecodedRevert({
      name: "Mixed",
      signature: "Mixed(uint256,address)",
      args: [100n, "0xabc"],
    })).toBe("Mixed(100, 0xabc)")
  })

  it("handles undefined", () => {
    expect(describeDecodedRevert(undefined)).toBe("unknown revert")
  })

  it("handles null", () => {
    expect(describeDecodedRevert(null)).toBe("unknown revert")
  })
})
